src/main/java/com/pinterest/secor/consumer/Consumer.java

```diff
@@ -71,7 +71,7 @@ private void init() throws Exception {
         FileRegistry fileRegistry = new FileRegistry(mConfig);
         UploadManager uploadManager = ReflectionUtil.createUploadManager(mConfig.getUploadManagerClass(), mConfig);
 
-        mUploader = new Uploader(mConfig, mOffsetTracker, fileRegistry, uploadManager);
+        mUploader = ReflectionUtil.createUploader(mConfig.getUploaderClass(), mConfig, mOffsetTracker, fileRegistry, uploadManager);
```

Comment:
@HenryCaiHaiying I was about to take that approach, however, other methods (e.g. createMessageTransformer) invoke the constructor with parameters. I would change it to newInstance + init anyway as you requested.